Set password hint on BSD, fill selection on macOS again

klemensn opened the pull request:
- Set x-kde-passwordManagerHint on BSD
- Hoist mime handling out of Q_OS_* macros
- Defer filling clipboard
- Simplify Q_OS_* macro logic
- Fill selction on macOS again

## Testing strategy
Tested on OpenBSD/amd64 7.2-current with `keepassxc-2.7.4p1-browser` and `copyq-6.3.2`.
This PR makes passwords copied from KeePassXC no longer appear in CopyQ (after adding a command that ignores values containing the password hint).

## Type of change
- ✅ Bug fix (non-breaking change that fixes an issue)
- ✅ Refactor (significant modification to existing code)

klemensn committed:
Hoist mime handling out of Q_OS_* macros

Both cases, i.e. MacOS and not-MacOS, set the text up front.
Deduplicate to simplify.

klemensn committed:
Defer filling clipboard

Both cases, i.e. MacOS and not-MacOS, set fill the clipboard once MIME
data is set up.  Deduplicate to simplify.

klemensn committed:
Set x-kde-passwordManagerHint on BSD

Linux and BSD are the same in this regard.  Use `Q_OS_UNIX` to cover
both families.

Found while configuring CopyQ to ignore clipboard changes with this MIME
type as per their FAQ[0] and wondering why it had no effect on OpenBSD.

0: https://copyq.readthedocs.io/en/latest/faq.html#how-to-omit-storing-text-copied-from-specific-windows-like-a-password-manager

klemensn committed:
Simplify Q_OS_* macro logic

Nested `#ifdef` are easy to get wrong, especially when inner ones do
not indent (for example `# ifdef ONE_LEVEL_DEEP_MEANS_ONE_SPACE`).

Flatten out the three branches for MacOS, Windows and UNIX and wrap the
later MacOS specific code again.

`defined()` must be used here as `#elifdef` is not a thing and `#elif`
cannot be used as it requires the macro to be defined, which is not how
`Q_OS_*` macros work -- this little but important detailed got pointed
out by KeePassXC's CI tests after submitting this PR as draft.

klemensn commented:
I've split this into multiple small commits since they are a mix of pure style changes, line reordering and behaviour changes (for some operating systems).

There does not seem to be a single definitive style wrt. `Q_OS_*` macro handling, i.e. some palces use `#ifdef` while others prefer `#if defined()`.

klemensn commented:
Reshuffled such that the first commit contains the actual fix while the next/last three ones contain optional cleanup.
Just in case the latter is deemed not needed (I hope not).

Sorry for the noise.

droidmonkey commented:
We will merge all the commits anyway. But you need to reintroduce the Q_OS_LINUX check. Change the Unix check to also `|| defined(Q_OS_LINUX)`

droidmonkey commented on src/gui/Clipboard.cpp:
```diff
@@ -52,24 +52,21 @@ void Clipboard::setText(const QString& text, bool clear)
 #endif
-    clipboard->setMimeData(mime, QClipboard::Clipboard);
 
+#if !defined(Q_OS_MACOS)
```
Why was this introduced?

klemensn replied:
> Why was this introduced?

This is what the existing code already does:

```
#ifdef Q_OS_MACOS
...
#else
    if (clipboard->supportsSelection()) {
        clipboard->setMimeData(mime, QClipboard::Selection);
    }
#endif
```
https://github.com/keepassxreboot/keepassxc/blob/a6d3f973fa8449f0f7dac864b3bd3928c29c649f/src/gui/Clipboard.cpp#L55-L72

droidmonkey replied:
Interesting that seems like a bug to me?

klemensn replied:
> Interesting that seems like a bug to me?

I am not at all familiar with clipboard handling on MacOS or Windows, so my PR does not change their functionality.

If this is indeed something to fix, could we do so in another PR?

droidmonkey replied:
Yes it is a bug, we used to use "setText" now we properly use "setMimeData" so this code should apply to all platforms:

![image](https://user-images.githubusercontent.com/2809491/210265812-e5df70f8-4649-4dbd-b4ad-df0f7f5d536a.png)

https://github.com/keepassxreboot/keepassxc/commit/dc7322cc4a08b64e72bf19e3b48502db682a65f6#diff-410a4485502cc323cf953be073436e3326ba2903f4c1912e9872662ce2446687

klemensn replied:
> [dc7322c#diff-410a4485502cc323cf953be073436e3326ba2903f4c1912e9872662ce2446687](https://github.com/keepassxreboot/keepassxc/commit/dc7322cc4a08b64e72bf19e3b48502db682a65f6#diff-410a4485502cc323cf953be073436e3326ba2903f4c1912e9872662ce2446687)

So the newly added `#endif` in line 55 should've probably landed as line 52, between clipboard and selection handling?

The diff you show seems is from 2017, which implies that KeePassXC has been failing to set the selection on MacOS for the last five years?

Looking at it with the context provided, I can add a proper commit fixing this and expand the PR a little.
That should be simplest.

klemensn replied:
> Looking at it with the context provided, I can add a proper commit fixing this and expand the PR a little.

Done with 69e234b.

klemensn committed:
Fill selction on macOS again

dc7322cc4a08b64e72bf19e3b48502db682a65f6
"Improve macOS pasteboard handling" in 2017 accidentially made selection
handling !macOS only by moving `#ifdef Q_OS_MACOS`'s `#endif` too far.

`supportsSelection()` takes care of feature detection already, so use it
on macOS again, as intended.

Pointed out during review by Jonathan White (droidmonkey).

klemensn commented:
Ubuntu times out building with Clang, I'm confident this is unrelated to this PR:
https://ci.keepassxc.org/buildConfiguration/KeePassXC_UbuntuLinux/222515?hideProblemsFromDependencies=false&hideTestsFromDependencies=false&expandBuildChangesSection=true&expandBuildProblemsSection=true&showLog=222511_1735_1628&logFilter=debug&logView=flowAware

droidmonkey commented:
Yes please ignore

klemensn commented:
Did anyone give this a try on macOS, by any chance?
Is such a runtime test required to merge this?

droidmonkey commented:
This one is good, im going to merge prs this weekend
